src/python/zquantum/core/evolution.py

```diff
@@ -1,157 +1,250 @@
+        time: time of evolution
+    Returns:
+        Circuit: Circuit representing evolved pyquil term.
+    """
+    if isinstance(time, sympy.Expr):
+        circuit = Circuit(pyquil.paulis.exponentiate(term))
+        for gate in circuit.gates:
+            if len(gate.params) > 1:
+                raise (
+                    NotImplementedError(
+                        "Time evolution of multi-parametered gates with symbolic "
+                        "parameters is not supported."
+                    )
                 )
-
+            elif gate.name == "Rz" or gate.name == "PHASE":
+                # We only want to modify the parameter of Rz gate or PHASE gate.
+                gate.params[0] = gate.params[0] * time
     else:
-        raise ValueError("Currently the method {} is not supported".format(method))
+        exponent = term * time
+        assert isinstance(exponent, pyquil.paulis.PauliTerm)
+        circuit = Circuit(pyquil.paulis.exponentiate(exponent))
+    return circuit
+
+
+@time_evolution_for_term.register
+def _time_evolution_for_term_qubit_operator(
+    term: QubitOperator, time: Union[float, sympy.Expr]
+) -> Circuit:
+    """Evolves a Pauli term for a given time and returns a circuit representing it.
+    Args:
+        term: Pauli term to be evolved
+        time: time of evolution
+    Returns:
+        Circuit: Circuit representing evolved pyquil term.
+    """
+
+    if len(term.terms) != 1:
+        raise ValueError("This function works only on a single term.")
+    term_components = list(term.terms.keys())[0]
+    base_changes = []
+    base_reversals = []
+    cnot_gates = []
+    central_gate = None
+    term_types = [component[1] for component in term_components]
+    qubit_indices = [component[0] for component in term_components]
+    coefficient = list(term.terms.values())[0]
+
+    for i, (term_type, qubit_id) in enumerate(zip(term_types, qubit_indices)):
+        # TODO: comments
+        if term_type == "X":
+            base_changes.append(Gate("H", qubits=[Qubit(qubit_id)]))
+            base_reversals.append(Gate("H", qubits=[Qubit(qubit_id)]))
+        elif term_type == "Y":
+            base_changes.append(
+                Gate("Rx", qubits=[Qubit(qubit_id)], params=[np.pi / 2])
+            )
+            base_reversals.append(
+                Gate("Rx", qubits=[Qubit(qubit_id)], params=[-np.pi / 2])
+            )
+        if i == len(term_components) - 1:
+            central_gate = Gate(
+                "Rz", qubits=[Qubit(qubit_id)], params=[2 * time * coefficient]
+            )
+        else:
+            cnot_gates.append(
+                Gate("CNOT", qubits=[Qubit(qubit_id), Qubit(qubit_indices[i + 1])])
+            )
+
+    circuit = Circuit()
+    for gate in base_changes:
+        circuit.gates.append(gate)
+
+    for gate in cnot_gates:
+        circuit.gates.append(gate)
 
-    return output
+    circuit.gates.append(central_gate)
+
+    for gate in cnot_gates[::-1]:
```

Comment:
A more idiomatic way of looping in reverse is: `for gate in reversed(cnot_gates)`. It is also faster, as it generates items instead of creating a copy of the list which is what a list slice `[::-1]` does.